physical/file/file.go

```diff
@@ -247,10 +247,16 @@ func (b *FileBackend) ListInternal(prefix string) ([]string, error) {
 	}
 
 	for i, name := range names {
-		if name[0] == '_' {
-			names[i] = name[1:]
-		} else {
+		fi, err := os.Stat(filepath.Join(path, name))
+		if err != nil {
+			return nil, err
+		}
+		if fi.IsDir() {
 			names[i] = name + "/"
+		} else {
+			if name[0] == '_' {
```

Comment:
Why do we strip off underscores in this case?

Reply:
The way the file backend works right now is that it uses underscores to differentiate between a file and a directory since vault's storage can use the same name for both.

Which as I'm typing this makes me realize something, so stand by on the review until I fix something that this broke :-)

Reply:
Actually, I didn't break it, I just thought I might have. Added tests that show that it's working as intended.

Basically a file gets an underscore prepended to it so that the names on a filesystem don't collide between a same-named directory and file. (This does mean that this can still be defeated by e.g. creating a directory name with a double underscore matching another file that starts with a single underscore, but, c'est la vie). So during listing, because the user-facing paths shouldn't know about this, it gets stripped off. Since that underscore is only added to actual keys (files), and not to directories, we only strip off the underscore if what we're looking at is is a file. The earlier behavior would strip it off unilaterally, which was incorrect.